### ISSUE TITLE
Result of `renv::update(lock = TRUE)` differs from `renv::update(lock = FALSE); renv::snapshot()`

### ISSUE THREAD

**Issue by salim-b** [rstudio/renv#1876]
renv 1.0.6 introduced a `lock` parameter to  `renv::update()` via https://github.com/rstudio/renv/commit/3c136470935e32175bcfa401a5387591e469f083, implementing https://github.com/rstudio/renv/issues/1849 (thanks again!).

I just tested this new parameter for the first time and noticed it suffers from a similar issue as https://github.com/rstudio/renv/issues/1828:

When I run (in a project using renv's **explicit** snapshot type)

```r
renv::update(prompt = FALSE, lock = TRUE)
```

it produces the following output

```
- Checking for updated packages ... Done!
The following package(s) will be updated:

# CRAN -----------------------------------------------------------------------
- spatial   [7.3-15 -> 7.3-17]

The following package(s) will be installed:
- spatial [7.3-17]
These packages will be installed into "~/Arbeit/ZDA/Git/zdaarau/fokus_reports/input/renv/library/R-4.3/x86_64-pc-linux-gnu".

# Installing packages --------------------------------------------------------
- Installing spatial ...                        OK [linked from cache]
Successfully installed 1 package in 16 milliseconds.
- Lockfile written to "~/Arbeit/ZDA/Git/zdaarau/fokus_reports/input/renv/renv.lock".
```

and renv writes the following lines to my project's `renv.lock`

```json
    "spatial": {
      "Package": "spatial",
      "Version": "7.3-17",
      "Source": "Repository",
      "Repository": "RSPM",
      "Requirements": [
        "R",
        "graphics",
        "stats",
        "utils"
      ],
      "Hash": "1229a01b4ec059e9f2396724f2ec9010"
    },
```

although package `spatial` is neither a direct, nor indirect project dependency, i.e. it is not listed in my project's `DESCRIPTION` file, nor is it a required dependency of any other listed package. This results in the project being "out-of-sync" (until I manually revert the lockfile addition above). `renv::status()` at this point outputs

```
The following package(s) are in an inconsistent state:

 package installed recorded used
 spatial y         y        n   

See ?renv::status() for advice on resolving these issues.
```

---

When I OTOH run

```r
renv::update(prompt = FALSE)
renv::snapshot(prompt = FALSE)
```

it produces the following output (which is exactly the same as above)

```
- Checking for updated packages ... Done!
The following package(s) will be updated:

# CRAN -----------------------------------------------------------------------
- spatial   [7.3-15 -> 7.3-17]

The following package(s) will be installed:
- spatial [7.3-17]
These packages will be installed into "~/Arbeit/ZDA/Git/zdaarau/fokus_reports/input/renv/library/R-4.3/x86_64-pc-linux-gnu".

# Installing packages --------------------------------------------------------
- Installing spatial ...                        OK [linked from cache]
Successfully installed 1 package in 16 milliseconds.
```

```
- Lockfile written to "~/Arbeit/ZDA/Git/zdaarau/fokus_reports/input/renv/renv.lock".
```

but renv ***does not alter my project's lockfile*** and everything is fine afterwards.

**Comment by salim-b** [comment]
This issue has been alleviated by https://github.com/rstudio/renv/commit/447839181ac13e190699a9aa4eedd1d514262d0e, which is part of renv 1.1.3+.

The problem now only occurs if I set `all = TRUE`, i.e.:

```r
renv::update(prompt = FALSE,
             lock = TRUE,
             all = TRUE)
```

As hinted above, I think the issue is specific to renv's ***explicit*** snapshot type, i.e. renv adds packages from non-project libraries to the lockfile (`spatial` in the example above) regardless whether they're listed in the project's `DESCRIPTION` file or not.

**Comment by kevinushey** [comment]
Thanks for the bug report. I have a question on the behavior in this scenario... suppose you call:

```
renv::update("spatial", lock = TRUE)
```

but "spatial" does not actually appear to be a dependency in the project; that is, it's not declared in the DESCRIPTION file or (if implicit snapshot mode) used in any `.R` files. Would you expect "spatial" to enter the lockfile in this scenario, or would it make sense to re-generate the lockfile but just in the "regular" way as though `renv::snapshot()` were called?

**Comment by salim-b** [comment]
> Would you expect "spatial" to enter the lockfile in this scenario (...)

No, I would only expect `spatial` to the enter the lockfile in such a scenario if it would be a (transitive) dependency of another package that was actually declared in the `DESCRIPTION` file.

> would it make sense to re-generate the lockfile but just in the "regular" way as though `renv::snapshot()` were called?

I guess so? But I don't know enough about renv internals to give a substantiated answer here...

**Comment by kevinushey** [comment]
Sorry, to be more succinct: I wonder if `renv::update(lock = TRUE)` should just be identical to `renv::update(); renv::snapshot()`.

**Comment by salim-b** [comment]
Ah, ok, now I understand. 

> I wonder if `renv::update(lock = TRUE)` should just be identical to `renv::update(); renv::snapshot()`.

This would be totally fine with me, I don't have a usecase where this would interfere.

> Or, does there need to be some filter saying that "only packages which were updated (and maybe their dependencies) should be updated in the lockfile, and only if those packages are already listed as dependencies"...

There might be users wanting that additional restraint/filtering, I dunno. I have a hard time coming up with a situation where you'd want to update a set of pkgs, but at the same time only want to have a *sub*set of those updates be reflected in the lockfile... maybe because of messy workflows this could be confusing for certain users, but I don't think it could become a serious problem.

**Comment by benlake** [comment]
> I wonder if renv::update(lock = TRUE) should just be identical to renv::update(); renv::snapshot().

Coming from experience as a user of other language package managers, yes, those definitely should result in the same outcome. I actually find the default segregation of install/update from the updating of the lock file alarming. I hope `snapshot` fades out in favor of defaulting `lock=TRUE` for install/update.

Though, the behavior of capturing packages that are "in use" probably plays a part, and in my opinion, that too needs to be disconnected from snapshot-ing. Definitely help me understand what packages I have that I might not need, but don't prevent me from specifying packages I want because I'm not _currently_ using them.

**Comment by kevinushey** [comment]
> Though, the behavior of capturing packages that are "in use" probably plays a part, and in my opinion, that too needs to be disconnected from snapshot-ing. Definitely help me understand what packages I have that I might not need, but don't prevent me from specifying packages I want because I'm not currently using them.

I think you're more-or-less asking for "explicit" snapshots, which gives a way for users to explicitly enumerate their project's dependency in the DESCRIPTION file -- this is discussed in https://rstudio.github.io/renv/reference/snapshot.html#snapshot-types.

**Comment by benlake** [comment]
@kevinushey Ah, I hadn't looked on the other side of the workflow (snapshot vs. install) to see if there was a more explicit path. Appreciate the call out, and that is certainly more inline with what I'd desire. Though, having an entry in DESCRIPTION is not the same as getting it into the lock file, where `restore` can do its thing. A particular pain point for me at the moment is the requirement that R packages, including their OS dependencies, must be installed for `snapshot()` to function, regardless of the "type". That said, while reviewing the `snapshot` options, I bumped into `record`, which does provide a means of getting entries into the lock file without requiring package installation. The nuance I alluded to before was a desire for the version lookup of `install_packages` without the actual installation, and the ability to update the lockfile with the discovered version which is provided by `record`.

I don't want to derail this issue more, so I'll stop there. Appreciate the response on my tangent! :)

**Comment by kevinushey** [comment]
Thanks! Just to close that loop, if you want to directly add or modify records in the lockfile, you might find `renv::checkout()` (use whatever packages are available on CRAN / your configured R package repositories), or `renv::record()` (add specific packages, using the remote to declare that package). We can discuss this further in a separate thread if you have more comments.
